### PR TITLE
rlp, p2p, p2p/discover: EIP-8 changes

### PR DESCRIPTION
**Pull request by fjl**
Implements https://github.com/ethereum/EIPs/pull/49

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Thu Dec 24 01:05:49 UTC 2015_

**Comment by fjl** [comment]
@obscuren @karalabe @Gustav-Simonsson please review

File: crypto/ecies/ecies.go [hunk 1]
```diff
@@ -192,11 +192,9 @@ func concatKDF(hash hash.Hash, z, s1 []byte, kdLen int) (k []byte, err error) {
 // messageTag computes the MAC of a message (called the tag) as per
 // SEC 1, 3.5.
 func messageTag(hash func() hash.Hash, km, msg, shared []byte) []byte {
-	if shared == nil {
-		shared = make([]byte, 0)
-	}
```
Comment (karalabe) [inline crypto/ecies/ecies.go]: Was this `shared` thing never used until now? Why is it only added to the hash now?

File: rlp/decode.go [hunk 2]
```diff
@@ -63,11 +63,11 @@ type Decoder interface {
...
 //
-// The decoding of struct fields honours one particular struct tag,
-// "nil". This tag applies to pointer-typed fields and changes the
+// The decoding of struct fields honours two struct tags, "nil" and
+// "..". The "nil" tag applies to pointer-typed fields and changes the
 // decoding rules for the field such that input values of size zero
-// decode as a nil pointer. This tag can be useful when decoding recursive
-// types.
+// decode as a nil pointer. This tag can be useful when decoding
+// recursive types.
```
Comment (karalabe) [inline rlp/decode.go]: Since you explain here how the `nil` tag works, please also add how the `..` works.

File: rlp/decode_test.go [hunk 3]
```diff
@@ -437,6 +452,23 @@ var decodeTests = []decodeTest{
...
+		ptr:   new(invalidDotDot2),
+		error: "rlp: invalid struct tag \"..\" for rlp.invalidDotDot2.B (field type is not slice)",
+	},
+
+	// struct tag ".."
+	{
+		input: "C3010203",
+		ptr:   new(dotdotRaw),
+		value: dotdotRaw{A: 1, DotDot: []RawValue{unhex("02"), unhex("03")}},
+	},
```
Comment (karalabe) [inline rlp/decode_test.go]: Please add a test case where the tag is correctly set, but the input RLP contains something wrong. E.g. you want to parse into a list of strings and one of the elements it an int, not a string. Or something like this, to verify that mixing elements are caught and does not crash or similar.

File: crypto/ecies/ecies.go (hunk 1 above)
Comment (fjl) [inline crypto/ecies/ecies.go]: We didn't use it until now.

File: rlp/decode_test.go (hunk 3 above)
Comment (karalabe) [inline rlp/decode_test.go]: Also it would be nice to see examples where:

 * `..` is used, but only one element is supplied
 * `..` is used, but 0 elements are supplied

Comment (fjl) [inline rlp/decode_test.go]: The example already checks this.

Comment (karalabe) [inline rlp/decode_test.go]: I'm not seeing either of the three cases checked?

File: rlp/encode.go [hunk 4]
```diff
@@ -507,20 +507,21 @@ func writeInterface(val reflect.Value, w *encbuf) error {
...
 }
 
-func makeSliceWriter(typ reflect.Type) (writer, error) {
+func makeSliceWriter(typ reflect.Type, ts tags) (writer, error) {
 	etypeinfo, err := cachedTypeInfo1(typ.Elem(), tags{})
 	if err != nil {
 		return nil, err
 	}
 	writer := func(val reflect.Value, w *encbuf) error {
-		lh := w.list()
```
Comment (karalabe) [inline rlp/encode.go]: Where to you start the list in the new code? I can only see the `listEnd` being used.

File: rlp/typecache.go [hunk 5]
```diff
@@ -101,7 +117,10 @@ func structFields(typ reflect.Type) (fields []field, err error) {
 }
 
 func parseStructTag(tag string) tags {
-	return tags{nilOK: tag == "nil"}
+	return tags{
+		nilOK:  strings.Contains(tag, "nil"),
+		dotdot: strings.Contains(tag, ".."),
+	}
```
Comment (karalabe) [inline rlp/typecache.go]: I think it would catch potential errors if these would do strict checks and not `contains`. I.e. split the `tag` by commas, and check each element for `nil`, or `...` and also error out if there's something else in there.

File: rlp/encode_test.go [hunk 6]
```diff
@@ -214,6 +214,7 @@ var encTests = []encTest{
 	{val: simplestruct{A: 3, B: "foo"}, output: "C50383666F6F"},
 	{val: &recstruct{5, nil}, output: "C205C0"},
 	{val: &recstruct{5, &recstruct{4, &recstruct{3, nil}}}, output: "C605C404C203C0"},
+	{val: &dotdotRaw{A: 1, DotDot: []RawValue{unhex("02"), unhex("03")}}, output: "C3010203"},
```
Comment (karalabe) [inline rlp/encode_test.go]: Could you also add a case here when you only have one element and when zero elements? :D I would really like these corner cases covered by tests.

File: rlp/encode.go [hunk 7]
```diff
@@ -507,20 +507,21 @@ func writeInterface(val reflect.Value, w *encbuf) error {
...
+		if !ts.dotdot {
+			defer w.listEnd(w.list())
```
Comment (obscuren) [inline rlp/encode.go]: @karalabe here. `w.list()` is executed prior to the rest of the code.

File: rlp/typecache.go (hunk 5 above)
Comment (obscuren) [inline rlp/typecache.go]: What kind of potential errors?

**Comment by obscuren** [comment]
As discussed on gitter change the `..` to `tail` to make it more descriptive what the tag does.

File: p2p/discover/udp.go [hunk 8]
```diff
@@ -460,6 +468,29 @@ func isTemporaryError(err error) bool {
 	return ok && tempErr.Temporary() || isPacketTooBig(err)
 }
 
+func encodePacket1(priv *ecdsa.PrivateKey, ptype byte, req interface{}, additional []byte) ([]byte, error) {
```
Comment (karalabe) [inline p2p/discover/udp.go]: What does this method do? I don't see it used in any of the code.

File: p2p/rlpx.go [hunk 9]
```diff
@@ -234,7 +256,9 @@ func (h *encHandshake) secrets(auth, authResp []byte) (secrets, error) {
 	return s, nil
 }
 
-func (h *encHandshake) ecdhShared(prv *ecdsa.PrivateKey) ([]byte, error) {
+// ecdhShared returns the static shared secret, the result
```
Comment (karalabe) [inline p2p/rlpx.go]: Method name in the doc is wrong.

File: p2p/rlpx.go [hunk 10]
```diff
@@ -242,106 +266,77 @@ func (h *encHandshake) ecdhShared(prv *ecdsa.PrivateKey) ([]byte, error) {
...
 func initiatorEncHandshake(conn io.ReadWriter, prv *ecdsa.PrivateKey, remoteID discover.NodeID, token []byte) (s secrets, err error) {
-	h, err := newInitiatorHandshake(remoteID)
+	h := &encHandshake{initiator: true, remoteID: remoteID}
+	authMsg, err := h.makeAuthMsg(prv, token)
 	if err != nil {
 		return s, err
 	}
-	auth, err := h.authMsg(prv, token)
+	var authPacket []byte
+	if os.Getenv("RLPX_EIP8") != "" {
```
Comment (karalabe) [inline p2p/rlpx.go]: Wasn't this only used for testing? Shouldn't EIP-8 be used always?

Comment (karalabe) [inline p2p/rlpx.go]: Hmmm, or is this here so we push out the change to first only accept EIP-8, but don't initiate it and when people upgraded then switch over to use it by everyone?

**Comment by karalabe** [comment]
Code looks solid to me. The crypto aspects I cannot verify, but given the thorough tests + the cross client interop experiments, I guess those parts are correct. Please address the issues I've raised and the it's a Go from my part.

File: p2p/rlpx.go (hunk 10 above)
Comment (fjl) [inline p2p/rlpx.go]: Yes, it's more like the latter comment. I used the env variable for testing with other clients.
Would be nice to ensure that the majority of the network will send EIP-8 soon-ish, maybe through
something like 

```go
if time.Now().After(switchoverDate) {
    // send EIP-8
}
```

**Comment by fjl** [comment]
PTAL

**Comment by karalabe** [comment]
LGTM :+1:

**Comment by obscuren** [comment]
:+1:

**Comment by karalabe** [comment]
@fjl Please squash.

**Comment by obscuren** [comment]
@fjl please re-open.
